text/2580-ptr-meta.md

```diff
@@ -0,0 +1,397 @@
+but tries to stay compatible with future proposals that do.
+
+
+# Background
+[background]: #background
+
+Typical high-level code doesn’t need to worry about fat pointers,
+a reference `&Foo` “just works” wether or not `Foo` is a DST.
+But unsafe code such as a custom collection library may want to access a fat pointer’s
+components separately.
+
+In Rust 1.11 we *removed* a [`std::raw::Repr`] trait and a [`std::raw::Slice`] type
+from the standard library.
+`Slice` could be `transmute`d to a `&[U]` or `&mut [U]` reference to a slice
+as it was guaranteed to have the same memory layout.
+This was replaced with more specific and less wildly unsafe
+`std::slice::from_raw_parts` and `std::slice::from_raw_parts_mut` functions,
+together with `as_ptr` and `len` methods that extract each fat pointer component separatly.
+
+For trait objects, where we still have an unstable `std::raw::TraitObject` type
+that can only be used with `transmute`:
+
+```rust
+#[repr(C)]
+pub struct TraitObject {
+    pub data: *mut (),
+    pub vtable: *mut (),
+}
+```
+
+[`std::raw::Repr`]: https://doc.rust-lang.org/1.10.0/std/raw/trait.Repr.html
+[`std::raw::Slice`]: https://doc.rust-lang.org/1.10.0/std/raw/struct.Slice.html
+[`std::raw::TraitObjet`]: https://doc.rust-lang.org/1.30.0/std/raw/struct.TraitObject.html
+
+
+# Motivation
+[motivation]: #motivation
+
+We now have APIs in Stable Rust to let unsafe code freely and reliably manipulate slices,
+accessing the separate components of a fat pointers and then re-assembling them.
+However `std::raw::TraitObject` is still unstable,
+but it’s probably not the style of API that we’ll want to stabilize
+as it encourages dangerous `transmute` calls.
+This is a “hole” in available APIs to manipulate existing Rust types.
+
+For example [this library][lib] stores multiple trait objects of varying size
+in contiguous memory together with their vtable pointers,
+and during iteration recreates fat pointers from separate data and vtable pointers.
+
+The new `Thin` trait alias also expanding to [extern types] some APIs
+that were unnecessarily restricted to `Sized` types
+because there was previously no way to express pointer-thinness in generic code.
+
+[lib]: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=bbeecccc025f5a7a0ad06086678e13f3
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+Let’s build generic type similar to `Box<dyn Trait>`,
+but where the vtable pointer is stored in heap memory next to the value
+so that the pointer is thin.
+First, let’s get some boilerplate out of the way:
+
+```rust
+use std::marker::{PhantomData, Unsize};
+use std::ptr::{self, DynMetadata};
+
+trait DynTrait<Dyn> = Pointee<Metadata=DynMetadata<Dyn>>;
+
+pub struct ThinBox<Dyn: ?Sized + DynTrait<Dyn>> {
+    ptr: ptr::NonNull<WithMeta<()>>,
+    phantom: PhantomData<Dyn>,
+}
+
+#[repr(C)]
+struct WithMeta<T: ?Sized> {
+    vtable: DynMetadata,
```

Comment:
Doesn't `DynMetadata` need a generic parameter?

Reply:
Right, I forget to update this example when making that change.